File: src/ragrep/eval/harness.py
"""Evaluation harness: measures cross-source retrieval quality per stage.

Uses the entity graph to generate ground truth automatically — no manual
curation needed. For each cross-source entity (Jira ticket, service name),
checks whether retrieval surfaces related chunks from multiple sources.

Metrics per stage (dense, BM25, RRF, rerank):
  - source_recall: fraction of expected sources in top-K
  - entity_recall: fraction of ground truth chunks in top-K
  - mrr: reciprocal rank of first ground truth hit
"""

import json
import logging
import time
from dataclasses import dataclass, field
from pathlib import Path

import numpy as np

from ragrep.config import Config
from ragrep.eval.entity_graph import EntityGraph, EvalCase
from ragrep.ingest.embed import make_embedder
from ragrep.ingest.store import load_index
from ragrep.query.rerank import make_reranker
from ragrep.query.retrieve import _reciprocal_rank_fusion, bm25_search, dense_search

log = logging.getLogger(__name__)


@dataclass
class StageMetrics:
    """Metrics for a single retrieval stage on one eval case."""

    source_recall: float = 0.0  # fraction of expected sources in results
    entity_recall: float = 0.0  # fraction of ground truth chunks in results
    mrr: float = 0.0  # reciprocal rank of first ground truth hit
    sources_found: set[str] = field(default_factory=set)
    n_hits: int = 0


@dataclass
class CaseResult:
    """Eval result for one case across all stages."""

    entity_key: str
    query: str
    expected_sources: set[str]
    n_ground_truth: int
    stages: dict[str, StageMetrics] = field(default_factory=dict)


def _score_stage(
    result_indices: list[int],
    ground_truth: set[int],
    expected_sources: set[str],
    chunks: list,
    top_k: int,
) -> StageMetrics:
    """Score a retrieval stage's results against ground truth."""
    top = result_indices[:top_k]

    # Source recall
    found_sources = {chunks[i].source for i in top if i < len(chunks)}
    source_hits = found_sources & expected_sources
    source_recall = len(source_hits) / len(expected_sources) if expected_sources else 0

    # Entity recall
    gt_hits = [i for i in top if i in ground_truth]
    entity_recall = len(gt_hits) / len(ground_truth) if ground_truth else 0

    # MRR
    mrr = 0.0
    for rank, idx in enumerate(top, 1):
        if idx in ground_truth:
            mrr = 1.0 / rank
            break

    return StageMetrics(
        source_recall=source_recall,
        entity_recall=entity_recall,
        mrr=mrr,
        sources_found=source_hits,
        n_hits=len(gt_hits),
    )


def evaluate(config: Config, output_path: Path | None = None) -> None:
    """Run cross-source retrieval evaluation."""
    rc = config.retrieval

    # Load index
    log.info("Loading index from %s", config.index_dir)
    faiss_index, chunks, bm25 = load_index(config.index_dir)

    # Build entity graph
    log.info("Building entity graph from %d chunks", len(chunks))
    graph = EntityGraph.build(chunks)
    cross = graph.cross_source_entities()
    log.info("Found %d cross-source entities", len(cross))

    # Generate eval cases
    cases = graph.eval_cases(max_cases=50)
    if not cases:
        log.warning("No eval cases generated")
        return
    log.info("Generated %d eval cases", len(cases))

    # Load models
    embedder = make_embedder(config.embedding.provider, config.embedding.model_name, config.embedding.device)
    reranker = make_reranker(config.reranker.provider, config.reranker.model_name, config.reranker.device)

    # Run eval
    results: list[CaseResult] = []
    t0 = time.monotonic()

    for i, case in enumerate(cases, 1):
        result = _eval_case(
            case, embedder, reranker, faiss_index, bm25, chunks,
            top_k_dense=rc.top_k_dense, top_k_bm25=rc.top_k_bm25, rrf_k=rc.rrf_k,
            top_k_rerank=rc.top_k_rerank,
        )
        results.append(result)

        # Print per-case summary
        rrf = result.stages.get("rrf", StageMetrics())
        rerank = result.stages.get("rerank", StageMetrics())
        print(
            f"  [{i:2d}] src={rerank.source_recall:.0%} "
            f"entity={rerank.entity_recall:.0%} "
            f"mrr={rerank.mrr:.2f} "
            f"({rrf.n_hits}→{rerank.n_hits} hits) "
            f"— {case.entity_key}: {case.query[:60]}"
        )

    elapsed = time.monotonic() - t0

    # Summary
    _print_summary(results, elapsed)

    if output_path:
        _save_results(results, output_path)


def _eval_case(
    case: EvalCase,
    embedder: object,
    reranker: object,
    faiss_index: object,
    bm25: object,
    chunks: list,
    top_k_dense: int,
    top_k_bm25: int,
    rrf_k: int,
    top_k_rerank: int,
) -> CaseResult:
    """Evaluate a single case across all retrieval stages."""
    gt = case.ground_truth_indices
    expected = case.expected_sources
    top_k_eval = top_k_dense + top_k_bm25  # eval window for pre-rerank stages

    result = CaseResult(
        entity_key=case.entity_key,
        query=case.query,
        expected_sources=expected,
        n_ground_truth=len(gt),
    )

    # Embed query
    query_emb = embedder.embed_query(case.query)

    # Dense
    dense = dense_search(query_emb, faiss_index, chunks, top_k_dense)
    dense_indices = [idx for idx, _ in dense]
    result.stages["dense"] = _score_stage(dense_indices, gt, expected, chunks, top_k_dense)

    # BM25
    bm25_results = bm25_search(case.query, bm25, chunks, top_k_bm25)
    bm25_indices = [idx for idx, _ in bm25_results]
    result.stages["bm25"] = _score_stage(bm25_indices, gt, expected, chunks, top_k_bm25)

    # RRF
    fused = _reciprocal_rank_fusion(dense, bm25_results, rrf_k)
    rrf_indices = [idx for idx, _, _, _ in fused]
    result.stages["rrf"] = _score_stage(rrf_indices, gt, expected, chunks, top_k_eval)

    # Rerank (on top of RRF pool)
    from ragrep.models import SearchResult

    # Map chunk_id → chunk index for ground truth matching after rerank
    id_to_idx = {chunks[idx].id: idx for idx, _, _, _ in fused[:top_k_eval]}

    candidates = []
    for idx, d_score, b_score, rrf_score in fused[:top_k_eval]:
        chunk = chunks[idx]
        candidates.append(SearchResult(
            chunk_id=chunk.id,
            content=chunk.content,
            title=chunk.title,
            source=chunk.source,
            metadata=chunk.metadata,
            dense_score=d_score,
            bm25_score=b_score,
            rrf_score=rrf_score,
        ))

    reranked = reranker.rerank(case.query, candidates, top_k_rerank)
    rerank_indices = [id_to_idx[r.chunk_id] for r in reranked]
    result.stages["rerank"] = _score_stage(rerank_indices, gt, expected, chunks, top_k_rerank)

    return result


def _print_summary(results: list[CaseResult], elapsed: float) -> None:
    """Print aggregate metrics per stage."""
    stages = ["dense", "bm25", "rrf", "rerank"]

    print(f"\n{'Stage':>8s}  {'SrcRecall':>10s}  {'EntityRecall':>13s}  {'MRR':>6s}")
    print("-" * 45)

    for stage in stages:
        metrics = [r.stages.get(stage, StageMetrics()) for r in results]
        avg_src = np.mean([m.source_recall for m in metrics])
        avg_ent = np.mean([m.entity_recall for m in metrics])
        avg_mrr = np.mean([m.mrr for m in metrics])
        print(f"  {stage:>6s}  {avg_src:>10.1%}  {avg_ent:>13.1%}  {avg_mrr:>6.3f}")

    print(f"\n{len(results)} cases evaluated in {elapsed:.1f}s")


def _save_results(results: list[CaseResult], path: Path) -> None:
    """Save results as JSON for further analysis."""
    path.parent.mkdir(parents=True, exist_ok=True)

    data = []
    for r in results:
        case_data = {
            "entity_key": r.entity_key,
            "query": r.query,
            "expected_sources": sorted(r.expected_sources),
            "n_ground_truth": r.n_ground_truth,
            "stages": {},
        }
        for stage, m in r.stages.items():
            case_data["stages"][stage] = {
                "source_recall": round(m.source_recall, 3),
                "entity_recall": round(m.entity_recall, 3),
                "mrr": round(m.mrr, 3),
                "sources_found": sorted(m.sources_found),
                "n_hits": m.n_hits,
            }
        data.append(case_data)

    with open(path, "w") as f:
        json.dump(data, f, indent=2)
    print(f"Results saved to {path}")

File: rust/src/eval/harness.rs
//! Evaluation harness: measures cross-source retrieval quality per stage.

use std::collections::BTreeSet;
use std::path::Path;
use std::time::Instant;

use anyhow::{bail, Context, Result};
use serde::Serialize;

use crate::config::Config;
use crate::embed;
use crate::eval::entity::{EntityGraph, EvalCase};
use crate::index::{bm25::Bm25, flat::Flat, store};
use crate::models::Chunk;
use crate::query::{self, retrieve};

const MAX_CASES: usize = 50;

#[derive(Debug, Clone, Default, Serialize)]
pub struct StageMetrics {
    pub source_recall: f64,
    pub entity_recall: f64,
    pub mrr: f64,
    pub sources_found: BTreeSet<String>,
    pub n_hits: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct CaseResult {
    pub entity_key: String,
    pub query: String,
    pub expected_sources: BTreeSet<String>,
    pub n_ground_truth: usize,
    pub stages: std::collections::BTreeMap<String, StageMetrics>,
}

#[derive(Debug, Clone, Serialize)]
pub struct EvalSummary {
    pub cases: usize,
    pub elapsed_s: f64,
    pub results: Vec<CaseResult>,
}

pub fn evaluate(config: &Config, output_path: Option<&Path>) -> Result<EvalSummary> {
    let rc = &config.retrieval;
    let index_dir = config.index_dir();
    if !store::index_exists(&index_dir) {
        bail!(
            "no complete Rust index found at {} (run `ragrep ingest` first)",
            index_dir.display()
        );
    }

    eprintln!("Loading index from {}", index_dir.display());
    let embedder = embed::make(&config.embedding.provider, &config.embedding.model_name)?;
    let chunks = store::load_chunks(&index_dir)?;
    let flat = store::load_flat(&index_dir, embedder.dim())?;
    let bm25 = store::load_bm25(&index_dir)?;

    eprintln!("Building entity graph from {} chunks", chunks.len());
    let graph = EntityGraph::build(&chunks);
    let cross_count = graph.cross_source_entities(2).len();
    eprintln!("Found {cross_count} cross-source entities");

    let cases = graph.eval_cases(2, MAX_CASES);
    if cases.is_empty() {
        eprintln!("No eval cases generated");
        let summary = EvalSummary {
            cases: 0,
            elapsed_s: 0.0,
            results: Vec::new(),
        };
        if let Some(path) = output_path {
            save_results(&summary, path)?;
        }
        return Ok(summary);
    }
    eprintln!("Generated {} eval cases", cases.len());

    let start = Instant::now();
    let mut results = Vec::with_capacity(cases.len());
    for (i, case) in cases.iter().enumerate() {
        let result = eval_case(
            case,
            embedder.as_ref(),
            &flat,
            &bm25,
            &chunks,
            rc.top_k_dense,
            rc.top_k_bm25,
            rc.rrf_k,
            rc.top_k_rerank,
            &config.reranker.provider,
            &config.reranker.model_name,
        )?;
        let rrf = result.stages.get("rrf").cloned().unwrap_or_default();
        let rerank = result.stages.get("rerank").cloned().unwrap_or_default();
        println!(
            "  [{:2}] src={:.0}% entity={:.0}% mrr={:.2} ({}→{} hits) — {}: {}",
            i + 1,
            rerank.source_recall * 100.0,
            rerank.entity_recall * 100.0,
            rerank.mrr,
            rrf.n_hits,
            rerank.n_hits,
            case.entity_key,
            truncate(&case.query, 60),
        );
        results.push(result);
    }

    let elapsed_s = start.elapsed().as_secs_f64();
    print_summary(&results, elapsed_s);

    let summary = EvalSummary {
        cases: results.len(),
        elapsed_s,
        results,
    };
    if let Some(path) = output_path {
        save_results(&summary, path)?;
    }
    Ok(summary)
}

#[allow(clippy::too_many_arguments)]
fn eval_case(
    case: &EvalCase,
    embedder: &dyn embed::Embedder,
    flat: &Flat,
    bm25: &Bm25,
    chunks: &[Chunk],
    top_k_dense: usize,
    top_k_bm25: usize,
    rrf_k: usize,
    top_k_rerank: usize,
    rerank_provider: &str,
    rerank_model: &str,
) -> Result<CaseResult> {
    let top_k_eval = top_k_dense + top_k_bm25;
    let mut result = CaseResult {
        entity_key: case.entity_key.clone(),
        query: case.query.clone(),
        expected_sources: case.expected_sources.clone(),
        n_ground_truth: case.ground_truth_indices.len(),
        stages: std::collections::BTreeMap::new(),
    };

    let query_emb = embedder.embed_query(&case.query)?;
    let filters = query::Filters::default();

    let dense = retrieve::dense(flat, &query_emb, chunks, top_k_dense, &filters);
    let dense_indices: Vec<usize> = dense.iter().map(|(idx, _)| *idx).collect();
    result.stages.insert(
        "dense".to_string(),
        score_stage(
            &dense_indices,
            &case.ground_truth_indices,
            &case.expected_sources,
            chunks,
            top_k_dense,
        ),
    );

    let bm25_results = retrieve::bm25(bm25, chunks, &case.query, top_k_bm25, &filters);
    let bm25_indices: Vec<usize> = bm25_results.iter().map(|(idx, _)| *idx).collect();
    result.stages.insert(
        "bm25".to_string(),
        score_stage(
            &bm25_indices,
            &case.ground_truth_indices,
            &case.expected_sources,
            chunks,
            top_k_bm25,
        ),
    );

    let fused = retrieve::rrf(&dense, &bm25_results, rrf_k);
    let rrf_indices: Vec<usize> = fused.iter().map(|(idx, _, _, _)| *idx).collect();
    result.stages.insert(
        "rrf".to_string(),
        score_stage(
            &rrf_indices,
            &case.ground_truth_indices,
            &case.expected_sources,
            chunks,
            top_k_eval,
        ),
    );

    let pool_n = top_k_eval.min(fused.len());
    let pool = &fused[..pool_n];
    let docs: Vec<&str> = pool
        .iter()
        .map(|(idx, _, _, _)| chunks[*idx].content.as_str())
        .collect();
    let reranked = query::rerank::rerank(
        rerank_provider,
        rerank_model,
        &case.query,
        &docs,
        top_k_rerank,
    )?;
    let rerank_indices: Vec<usize> = reranked
        .into_iter()
        .filter_map(|item| pool.get(item.index).map(|(idx, _, _, _)| *idx))
        .collect();
    result.stages.insert(
        "rerank".to_string(),
        score_stage(
            &rerank_indices,
            &case.ground_truth_indices,
            &case.expected_sources,
            chunks,
            top_k_rerank,
        ),
    );

    Ok(result)
}

fn score_stage(
    result_indices: &[usize],
    ground_truth: &BTreeSet<usize>,
    expected_sources: &BTreeSet<String>,
    chunks: &[Chunk],
    top_k: usize,
) -> StageMetrics {
    let top = &result_indices[..result_indices.len().min(top_k)];

    let found_sources: BTreeSet<String> = top
        .iter()
        .filter_map(|idx| chunks.get(*idx))
        .map(|chunk| chunk.source.clone())
        .collect();
    let sources_found: BTreeSet<String> = found_sources
        .intersection(expected_sources)
        .cloned()
        .collect();

    let gt_hits: Vec<usize> = top
        .iter()
        .copied()
        .filter(|idx| ground_truth.contains(idx))
        .collect();

    let mrr = top
        .iter()
        .position(|idx| ground_truth.contains(idx))
        .map(|pos| 1.0 / (pos + 1) as f64)
        .unwrap_or(0.0);

    StageMetrics {
        source_recall: if expected_sources.is_empty() {
            0.0
        } else {
            sources_found.len() as f64 / expected_sources.len() as f64
        },
        entity_recall: if ground_truth.is_empty() {
            0.0
        } else {
            gt_hits.len() as f64 / ground_truth.len() as f64
        },
        mrr,
        sources_found,
        n_hits: gt_hits.len(),
    }
}

fn print_summary(results: &[CaseResult], elapsed_s: f64) {
    println!(
        "\n{:>8}  {:>10}  {:>13}  {:>6}",
        "Stage", "SrcRecall", "EntityRecall", "MRR"
    );
    println!("{}", "-".repeat(45));

    for stage in ["dense", "bm25", "rrf", "rerank"] {
        let metrics: Vec<StageMetrics> = results
            .iter()
            .map(|r| r.stages.get(stage).cloned().unwrap_or_default())
            .collect();
        let avg_src = mean(metrics.iter().map(|m| m.source_recall));
        let avg_ent = mean(metrics.iter().map(|m| m.entity_recall));
        let avg_mrr = mean(metrics.iter().map(|m| m.mrr));
        println!(
            "  {:>6}  {:>9.1}%  {:>12.1}%  {:>6.3}",
            stage,
            avg_src * 100.0,
            avg_ent * 100.0,
            avg_mrr
        );
    }

    println!("\n{} cases evaluated in {:.1}s", results.len(), elapsed_s);
}

fn mean(values: impl Iterator<Item = f64>) -> f64 {
    let mut total = 0.0;
    let mut count = 0usize;
    for value in values {
        total += value;
        count += 1;
    }
    if count == 0 {
        0.0
    } else {
        total / count as f64
    }
}

fn save_results(summary: &EvalSummary, path: &Path) -> Result<()> {
    if let Some(parent) = path.parent().filter(|p| !p.as_os_str().is_empty()) {
        std::fs::create_dir_all(parent)
            .with_context(|| format!("creating {}", parent.display()))?;
    }
    let raw = serde_json::to_string_pretty(summary).context("serializing eval results")?;
    std::fs::write(path, raw).with_context(|| format!("writing {}", path.display()))?;
    println!("Results saved to {}", path.display());
    Ok(())
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        return s.to_string();
    }
    let mut end = max;
    while end > 0 && !s.is_char_boundary(end) {
        end -= 1;
    }
    s[..end].to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{Chunk, MetaValue};
    use std::collections::BTreeMap;

    fn chunk(source: &str) -> Chunk {
        Chunk {
            id: source.to_string(),
            doc_id: source.to_string(),
            content: source.to_string(),
            title: source.to_string(),
            source: source.to_string(),
            metadata: BTreeMap::<String, MetaValue>::new(),
        }
    }

    #[test]
    fn score_stage_computes_source_entity_and_mrr() {
        let chunks = vec![chunk("git"), chunk("slack"), chunk("atlassian")];
        let expected_sources = ["git".to_string(), "atlassian".to_string()]
            .into_iter()
            .collect();
        let ground_truth = [0usize, 2usize].into_iter().collect();

        let metrics = score_stage(&[1, 2, 0], &ground_truth, &expected_sources, &chunks, 3);

        assert_eq!(metrics.source_recall, 1.0);
        assert_eq!(metrics.entity_recall, 1.0);
        assert_eq!(metrics.mrr, 0.5);
        assert_eq!(metrics.n_hits, 2);
    }

    #[test]
    fn score_stage_respects_top_k_window() {
        let chunks = vec![chunk("git"), chunk("atlassian")];
        let expected_sources = ["atlassian".to_string()].into_iter().collect();
        let ground_truth = [1usize].into_iter().collect();

        let metrics = score_stage(&[0, 1], &ground_truth, &expected_sources, &chunks, 1);

        assert_eq!(metrics.source_recall, 0.0);
        assert_eq!(metrics.entity_recall, 0.0);
        assert_eq!(metrics.mrr, 0.0);
    }
}
